internal/auth/password.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package auth
+
+import (
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/base64"
+	"fmt"
+	"io"
+
+	"github.com/juju/errors"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+// Password hides and protects a plain text passwords in Juju from accidentally
+// being consumed or printed to a log.
+type Password struct {
+	// password is the plain text password being protected
+	password []byte
+}
+
+const (
+	// defaultSaltLength is the default number of bytes that we randomly
+	// generate for new salts. The pbldf2 guidelines is to use 8 bytes for salt.
+	defaultSaltLength = 12
+
+	// pbkdf2Iterations is the number iterations to perform for pbkdf2. The
+	// more iterations performed the harder brute forcing is. Each iteration
+	// repeatedly hashes the password adding computational complexity.
+	//
+	// Warning updating this value will break already established passwords on
+	// the old value!
+	pbkdf2Iterations = 8192
+
+	// pbkdf2KeySize is the size of the produced key by pbkdf2KeySize, expressed
+	// in number of bytes.
+	//
+	// Warning updating this value will break already established passwords on
+	// the old value!
+	pbkdf2KeySize = 18
+)
+
+// HasPassword takes a password and corresponding salt to produce a hash of the
+// password. The resultant hash is safe for persistence and comparison.
+// If the sale provided to password hash is empty then a error satisfying
+// errors.NotValid is returned.
+func HashPassword(p Password, salt []byte) (string, error) {
+	if len(salt) == 0 {
```

Comment:
We should also limit the length of the size of a password to prevent resource starvation on the server. Maybe 1kb?

Reply:
Done & Done